Cannot use Scala 2.13 traits defining private case class

## Compiler version
 3.3.0-RC1-bin-20221128-8fba321-NIGHTLY
 3.2.0

## Minimized code

Needs minimization, I don't how it can be reproduced in the Vulpix
```Scala
//> using lib "net.liftweb:lift-json_2.13:3.5.0"

@main def Test = 
  println(net.liftweb.json.NoTypeHints)
```

Referenced code is in general structured as follows in Scala 2.13
```scala
trait TypeHints {
  val hints: List[Class[_]] = Nil
  def components: List[TypeHints] = List(this)

  def + (hints: TypeHints): TypeHints = CompositeTypeHints(components ::: hints.components)

  private[TypeHints] case class CompositeTypeHints(override val components: List[TypeHints]) extends TypeHints {
    val hints: List[Class[_]] = components.flatMap(_.hints)
  }
}

case object NoTypeHints extends TypeHints {
  val hints = Nil
}
```

## Output (click arrow to expand)
<details>

```scala
[error] java.lang.AssertionError: assertion failed: no companion CompositeTypeHints in dotty.tools.dotc.core.Scopes$MutableScope@52cd6990
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.core.Decorators$.assertingErrorsReported(Decorators.scala:279)
[error] dotty.tools.dotc.core.NamerOps$.findModuleBuddy(NamerOps.scala:73)
[error] dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$.setClassInfo(Scala2Unpickler.scala:91)
[error] dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler.parseToCompletion$1(Scala2Unpickler.scala:611)
[error] dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler.complete$$anonfun$1(Scala2Unpickler.scala:641)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler.atReadPos(Scala2Unpickler.scala:314)
[error] dotty.tools.dotc.core.unpickleScala2.Scala2Unpickler$LocalUnpickler.complete(Scala2Unpickler.scala:643)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:173)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:187)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:189)
[error] dotty.tools.dotc.core.DenotTransformers$InfoTransformer.transform(DenotTransformers.scala:42)
[error] dotty.tools.dotc.core.DenotTransformers$InfoTransformer.transform$(DenotTransformers.scala:34)
[error] dotty.tools.dotc.transform.FirstTransform.transform(FirstTransform.scala:37)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:833)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:120)
[error] dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:114)
[error] dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:107)
[error] dotty.tools.dotc.transform.ValueClasses$.isDerivedValueClass(ValueClasses.scala:17)
[error] dotty.tools.dotc.core.TypeErasure.dotty$tools$dotc$core$TypeErasure$$apply(TypeErasure.scala:595)
[error] dotty.tools.dotc.core.TypeErasure.eraseResult(TypeErasure.scala:799)
[error] dotty.tools.dotc.core.TypeErasure.eraseInfo(TypeErasure.scala:740)
[error] dotty.tools.dotc.core.TypeErasure$.transformInfo(TypeErasure.scala:244)
[error] dotty.tools.dotc.transform.Erasure.transform(Erasure.scala:97)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:833)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:120)
[error] dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:114)
[error] dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:107)
[error] dotty.tools.dotc.core.Symbols$.toDenot(Symbols.scala:497)
[error] dotty.tools.dotc.transform.Mixin.needsTraitSetter(Mixin.scala:174)
[error] dotty.tools.dotc.transform.Mixin.transformSym$$anonfun$1(Mixin.scala:153)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.Scopes$Scope.foreach(Scopes.scala:93)
[error] dotty.tools.dotc.transform.Mixin.transformSym(Mixin.scala:157)
[error] dotty.tools.dotc.core.DenotTransformers$SymTransformer.transform(DenotTransformers.scala:70)
[error] dotty.tools.dotc.core.DenotTransformers$SymTransformer.transform$(DenotTransformers.scala:65)
[error] dotty.tools.dotc.transform.Mixin.transform(Mixin.scala:113)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:833)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error] dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error] dotty.tools.dotc.core.Types$NamedType.computeDenot(Types.scala:2348)
[error] dotty.tools.dotc.core.Types$NamedType.denot(Types.scala:2311)
[error] dotty.tools.dotc.core.Types$NamedType.info(Types.scala:2299)
[error] dotty.tools.backend.jvm.BCodeHelpers.dotty$tools$backend$jvm$BCodeHelpers$$typeToTypeKind(BCodeHelpers.scala:813)
[error] dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind(BCodeHelpers.scala:285)
[error] dotty.tools.backend.jvm.BCodeHelpers$BCInnerClassGen.toTypeKind$(BCodeHelpers.scala:213)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.toTypeKind(BCodeSkelBuilder.scala:74)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.paramTKs$$anonfun$1(BCodeSkelBuilder.scala:99)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.paramTKs(BCodeSkelBuilder.scala:99)
[error] dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:802)
[error] dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:369)
[error] dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:290)
[error] dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:83)
[error] dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlockTo$$anonfun$1(BCodeBodyBuilder.scala:1052)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlockTo(BCodeBodyBuilder.scala:1052)
[error] dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:441)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:809)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:832)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:615)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen$$anonfun$1(BCodeSkelBuilder.scala:621)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:621)
[error] dotty.tools.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:233)
[error] dotty.tools.backend.jvm.GenBCodePipeline$Worker1.visit(GenBCode.scala:266)
[error] dotty.tools.backend.jvm.GenBCodePipeline$Worker1.run(GenBCode.scala:231)
[error] dotty.tools.backend.jvm.GenBCodePipeline.buildAndSendToDisk(GenBCode.scala:598)
[error] dotty.tools.backend.jvm.GenBCodePipeline.run(GenBCode.scala:564)
[error] dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:69)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:316)
[error] scala.collection.immutable.List.map(List.scala:250)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:320)
[error] dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:77)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:233)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1321)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:244)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:252)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:261)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:261)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:185)
[error] dotty.tools.dotc.Run.compile(Run.scala:169)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:193)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:248)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:183)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:211)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:179)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:177)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:463)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:418)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:506)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:405)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:171)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:534)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:488)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:425)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2363)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2313)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:30)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:30)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2311)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] java.base/java.lang.Thread.run(Thread.java:834)
[error]            
[error] stack trace is suppressed; run last rudder-core / Compile / compileIncremental for the full output
[error] (rudder-core / Compile / compileIncremental) java.lang.AssertionError: assertion failed: no companion CompositeTypeHints in dotty.tools.dotc.core.Scopes$MutableScope@52cd6990
```

```scala
Error: Unexpected error when compiling project_92a9a85014-c6c223d3e0: 'assertion failed: no companion CompositeTypeHints in Scope{
  def <init>(): Unit
  def hints: List[Class[?]]
  def hintFor(clazz: Class[?]): String
  def classFor(hint: String): Option[Class[?]]
  def containsHint_?(clazz: Class[?]): Boolean
  def deserialize: PartialFunction[(String, net.liftweb.json.JObject), Any]
  def serialize: PartialFunction[Any, net.liftweb.json.JObject]
  def components: List[net.liftweb.json.TypeHints]
  def +(hints: net.liftweb.json.TypeHints): net.liftweb.json.TypeHints
  case private[TypeHints] case class CompositeTypeHints?
  lazy module private[TypeHints] object
    net$liftweb$json$TypeHints$$CompositeTypeHints:
    TypeHints.this.CompositeTypeHints
  module private[TypeHints] object CompositeTypeHints?
}'
```
</details>

--- a/compiler/src/dotty/tools/dotc/core/NamerOps.scala
+++ b/compiler/src/dotty/tools/dotc/core/NamerOps.scala
@@ -67,11 +67,11 @@ object NamerOps:
       completer.withSourceModule(findModuleBuddy(name.sourceModuleName, scope))
 
   /** Find moduleClass/sourceModule in effective scope */
-  def findModuleBuddy(name: Name, scope: Scope)(using Context) = {
-    val it = scope.lookupAll(name).filter(_.is(Module))
-    if (it.hasNext) it.next()
+  def findModuleBuddy(name: Name, scope: Scope, alternate: Name = EmptyTermName)(using Context): Symbol =
+    var it = scope.lookupAll(name).filter(_.is(Module))
+    if !alternate.isEmpty then it ++= scope.lookupAll(alternate).filter(_.is(Module))
+    if it.hasNext then it.next()
     else NoSymbol.assertingErrorsReported(em"no companion $name in $scope")
-  }
 
   /** If a class has one of these flags, it does not get a constructor companion */
   private val NoConstructorProxyNeededFlags = Abstract | Trait | Case | Synthetic | Module | Invisible

--- a/compiler/src/dotty/tools/dotc/core/unpickleScala2/Scala2Unpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/unpickleScala2/Scala2Unpickler.scala
@@ -89,7 +89,11 @@ object Scala2Unpickler {
         val sourceModule = denot.sourceModule.orElse {
           // For non-toplevel modules, `sourceModule` won't be set when completing
           // the module class, we need to go find it ourselves.
-          NamerOps.findModuleBuddy(cls.name.sourceModuleName, denot.owner.info.decls)
+          val modName = cls.name.sourceModuleName
+          val alternate =
+            if cls.privateWithin.exists && cls.owner.is(Trait) then modName.expandedName(cls.owner)
+            else EmptyTermName
+          NamerOps.findModuleBuddy(modName, denot.owner.info.decls, alternate)
         }
         denot.owner.thisType.select(sourceModule)
       else selfInfo

--- a/tests/pos/i16443/Test_2.scala
+++ b/tests/pos/i16443/Test_2.scala
@@ -1,0 +1,2 @@
+@main def Test =
+  println(NoTypeHints)

--- a/tests/pos/i16443/TypeHints_1.scala
+++ b/tests/pos/i16443/TypeHints_1.scala
@@ -1,0 +1,15 @@
+// This should be run with Scala 2.13
+trait TypeHints {
+  val hints: List[Class[_]]
+  def components: List[TypeHints] = List(this)
+
+  def + (hints: TypeHints): TypeHints = CompositeTypeHints(components ::: hints.components)
+
+  private case class CompositeTypeHints(override val components: List[TypeHints]) extends TypeHints {
+    override val hints: List[Class[_]] = components.flatMap(_.hints)
+  }
+}
+
+case object NoTypeHints extends TypeHints {
+  override val hints = Nil
+}

